docs/src/ref/distributions.md

```diff
@@ -19,3 +17,207 @@ exponential
+   across multiple calls to this distribution (with the same arguments) are
+   `RND` expressions.
+
+**Importantly, Julia control flow constructs generally expect `CONST` values:
+the condition of an `if` or the range of a `for` loop cannot be `ARG` or
+`RND`.**
+_(Developer note: We could change this: it would not be too difficult, e.g., to
+add support for most use cases of `if`.)_
+
+The body expression as a whole must be a `RND` expression, representing a
+random variable. The behavior of the `@dist` definition is then to define a new
+distribution (with name `name`) that samples and evaluates the logpdf of the
+random variable represented by the `body` expression.
+
+Expressions are typed compositionally, with the following typing rules:
+
+1. **Literals and free variables are `CONST`s.** Literals and symbols that
+   appear free in the `@dist` body are of type `CONST`.
+
+2. **Arguments are `ARG`s.** Symbols bound as arguments in the `@dist`
+   declaration have type `ARG` in its body.
+
+3. **Drawing from a distribution gives `RND`.** If `d` is a distribution, and
+   `x_i` are of type `ARG` or `CONST`, `d(x_1, x_2, ...)` is of type `RND`.
+
+4. **Functions of `CONST`s are `CONST`s.** If `f` is a deterministic function
+   and `x_i` are all of type `CONST`, `f(x_1, x_2, ...)` is of type `CONST`.
+
+5. **Functions of `CONST`s and `ARG`s are `ARG`s.** If `f` is a
+   _differentiable_ function, and each `x_i` is either a `CONST` or a _scalar_
+   `ARG` (with at least one `x_i` being an `ARG`), then `f(x_1, x_2, ...)` is
+   of type `ARG`.
+
+6. **Functions of `CONST`s, `ARG`s, and `RND`s are `RND`s.** If `f` is one of a
+   special set of deterministic functions we've defined (`+`, `-`, `*`, `/`,
+   `exp`, `log`, `getindex`), and exactly one of its arguments `x_i` is of type
+   `RND`, then `f(x_1, x_2, ...)` is of type `RND`.
+
+One way to think about this, without all the rules, is that `CONST` values are
+"contaminated" by interaction with `ARG` values (becoming `ARG`s themselves),
+and both `CONST` and `ARG` are "contaminated" by interaction with `RND`.
+Thinking of the body as an AST, the journey from leaf node to root node always
+involves transitions in the direction of `CONST -> ARG -> RND`, never in
+reverse. (There are certain similarities here with [security
+types](https://en.wikipedia.org/wiki/Security_type_system), which also enforce
+a single direction of information flow -- anything that touches classified data
+is also classified. Also, Tabular v2's type system, which similarly separates
+deterministic and random values in its type system.)
+
+#### Restrictions
+Users may _not_ reassign to arguments (like `x` in the above example), and may
+not apply functions with side effects. Names bound to expressions of type `RND`
+must be used only once. e.g., `let x = normal(0, 1) in x + x` is not allowed.
+
+**TODO:** (Expand on scalar-vs.-vector problems here.)
```

Comment:
Is it OK to leave this TODO in the docs?

Reply:
Issue would be better -- filed https://github.com/probcomp/Gen/issues/200, thanks.